File: cp_rust/src/bin/cses-1618-trailing-zeros.rs
// Created by Ayush Biswas at 2025/07/10 12:05
// https://cses.fi/problemset/task/1618
use cp_lib::*;

// @code begin
use cpio::*;
use std::ops::DivAssign;

sol! {
    fn solution(
        n: usize,
    ) -> usize {
        (1..)
            .scan(n, |state, _| {
                state.div_assign(5);
                if *state > 0 {
                    Some(*state)
                } else {
                    None
                }
            })
            .sum()
    }
}

// @code end

File: cp_haskell/bin/cses-1618-trailing-zeros.hs
-- Created by Ayush Biswas at 2026/03/04 10:11
-- https://cses.fi/problemset/task/1618

-- @code begin

powInFact :: (Integral t) => t -> t -> t
powInFact 0 _ = 0
powInFact n i = n' + powInFact n' i
  where
    n' = n `div` i

main :: IO ()
main = do
  n :: Int <- read <$> getLine
  print $ powInFact n 5

-- @code end